src/main/java/com/opentable/kafka/util/OffsetMetrics.java

```diff
@@ -234,21 +269,85 @@ private void pollUnsafe(final String topic) {
                     );
         }
 
-        offsets.forEach((part, off) -> gauge(partitionName(topic, part, "offset")).set(off));
-        lag    .forEach((part, off) -> {
-            gauge(partitionName(topic, part, "lag")).set(off);
-            ((Histogram) metricMap.get(totalName(topic, "lag-distribution"))).update(off);
-        });
+        // The keys of offsets is a non-empty subset of the keys of sizes, and the keys of offsets and lags are
+        // identical.
+        final Set<Integer> knownParts = offsets.keySet();
+        final Set<Integer> unknownParts = Sets.difference(offsets.keySet(), knownParts);
+
+        subMap(knownParts, sizes).forEach((part, size) -> gauge(partitionName(topic, part, "size")).set(size));
+        unknownParts.forEach(              part        -> gauge(partitionName(topic, part, "size")).set(null));
+        gauge(totalName(topic, "size")).set(sumValues(subMap(knownParts, sizes)));
 
+        offsets     .forEach((part, off) -> gauge(partitionName(topic, part, "offset")).set(off));
+        unknownParts.forEach( part       -> gauge(partitionName(topic, part, "offset")).set(null));
         gauge(totalName(topic, "offset")).set(sumValues(offsets));
-        gauge(totalName(topic, "lag")).set(sumValues(lag));
+
+        lags.forEach((part, lag) -> {
+            gauge(partitionName(topic, part, "lag")).set(lag);
+            ((Histogram) metricMap.get(totalName(topic, "lag-distribution"))).update(lag);
+        });
+        unknownParts.forEach(part -> gauge(partitionName(topic, part, "lag")).set(null));
+        gauge(totalName(topic, "lag")).set(sumValues(lags));
     }
 
-    private AtomicLongGauge gauge(final String name) {
-        return (AtomicLongGauge) metricMap.get(name);
+    private LongGauge gauge(final String name) {
+        return (LongGauge) metricMap.get(name);
     }
 
     private static long sumValues(final Map<?, Long> map) {
-        return map.values().stream().mapToLong(size -> size).sum();
+        return map.values().stream().mapToLong(x -> x).sum();
+    }
+
+    /**
+     * Sub-map of {@code map} whose keys are the elements of {@code keys} with corresponding values from {@code map}.
+     */
+    private static <K, V> Map<K, V> subMap(final Collection<K> keys, final Map<K, V> map) {
```

Comment:
Is this equivalent to a construction like
```java
Maps.filterKeys(map, keys::contains)
```

Reply:
Yes.  Implemented in b25d714e1da435924a022b8a7b3f25c80ba116b2.

Reply:
(Thank you for the suggestion!!)